### ISSUE TITLE
Why does v0.6.0 query a custom index domain-name, instead of crates.io API ?

### ISSUE THREAD

**Issue by BarbossHack** [serayuzgur/crates#197]
### **Is your feature request related to a problem? Please describe.**
It is not possible to use "Crates" offline anymore (since v0.6.0), as it have to query https://api.crates-vsc.space

### **Describe the solution you'd like**
Instead of getting a crate version (for example base64) from https://api.crates-vsc.space/index/versions/base64 , it should query https://crates.io/api/v1/crates/base64/versions , you will have the same results (see the screenshot just bellow), and we will able to deploy our own offline crates.io API if needed
![image](https://github.com/serayuzgur/crates/assets/49869902/49294464-56be-4cff-ad7a-e9c42414116f)

Or at least, have the same api URI than crates.io (`/crates/NAME/versions` and not `/index/versions/NAME`), and same json response, to allow us to set our own offline api domain name

Thank you :+1:

**Comment by kristof-mattei** [comment]
I think this is actually quite concerning.

All of the sudden we have data being send to a 3rd party URL. There are no release notes, except for [`all cis migration done`](https://github.com/serayuzgur/crates/commit/9eb0ff659d3ddea4150cb48ceb8428ab5cf444c9), we don't know the owner and we don't know what happens with our data. 

The WHOIS shows a registrant in Iceland (IS):

https://www.whois.com/whois/crates-vsc.space

```
...
Registrant Organization: Privacy service provided by Withheld for Privacy ehf
Registrant State/Province: Capital Region
Registrant Country: IS
Registrant Email: Please query the RDDS service of the Registrar of Record identified in this output for information on how to contact the Registrant, Admin, or Tech contact of the queried domain name.
...
```

**Comment by BarbossHack** [comment]
Release notes are here https://github.com/serayuzgur/crates/blob/master/CHANGELOG.md, but yeah it wasn't done the usual way (not released on github), and I still  see no reason not to use the official API

**Comment by serayuzgur** [comment]
Okay, here's the whole deal: At first, we were using the crates.io API, but they asked us not to and switch to the GitHub API instead. So, we did that, but then people started complaining about API limits.

To solve that problem, we added local index support, thinking it'd make things better. But then folks started saying cargo wasn't getting updated, and we couldn't see the latest versions or have a local registry.

Eventually, cargo went through a big protocol change, which was a major bummer for us. So, I took matters into my own hands and created a similar API using the crates.io GitHub index.

Why did I do it? Well, managing that local index was driving me nuts, and I simply didn't have enough time to deal with all the user issues. Plus, that protocol change was a real dealbreaker. And don't even get me started on how huge the extension code was getting!

The only info we send to the server is the crate name you're requesting. Don't worry; it's all open source, and you can check out the data we send—it's transparent like that. I got the cheapest domain because, honestly, nobody's paying or donating, and the privacy option that came with it was just the default, so I didn't bother changing it or thinking too much about it.

**Comment by BarbossHack** [comment]
Thank you for your explanations @serayuzgur :+1: 

**Quote from my first comment :** 

> Or at least, have the same api URI than crates.io (/crates/NAME/versions and not /index/versions/NAME), and same json response, to allow us to set our own offline api domain name

Should it be possible ?

Thank for your work !


> At first, we were using the crates.io API, but they asked us not to and switch to the GitHub API instead.

wtf are they thinking at rust-lang ...

**Comment by BarbossHack** [comment]
or should I open a new issue ?

**Comment by BarbossHack** [comment]
(Also if you publish your server api on github I would be able to make a PR to help you)

**Comment by kristof-mattei** [comment]
> Okay, here's the whole deal: At first, we were using the crates.io API, but they asked us not to and switch to the GitHub API instead. So, we did that, but then people started complaining about API limits.
> 
> 
> 
> To solve that problem, we added local index support, thinking it'd make things better. But then folks started saying cargo wasn't getting updated, and we couldn't see the latest versions or have a local registry.
> 
> 
> 
> Eventually, cargo went through a big protocol change, which was a major bummer for us. So, I took matters into my own hands and created a similar API using the crates.io GitHub index.
> 
> 
> 
> Why did I do it? Well, managing that local index was driving me nuts, and I simply didn't have enough time to deal with all the user issues. Plus, that protocol change was a real dealbreaker. And don't even get me started on how huge the extension code was getting!
> 
> 
> 
> The only info we send to the server is the crate name you're requesting. Don't worry; it's all open source, and you can check out the data we send—it's transparent like that. I got the cheapest domain because, honestly, nobody's paying or donating, and the privacy option that came with it was just the default, so I didn't bother changing it or thinking too much about it.
> 
> 
> 
> 

I appreciate the update. 

With all of that said and crates.io now being sparse, is this still an issue?

**Comment by serayuzgur** [comment]
> or should I open a new issue ?

No, there's no need for that. I don't plan to make them identical. You see, the crates.io API isn't quite designed for the specific task of version listing; it generates a response JSON that's quite bloated with unnecessary data. Given this situation, I've decided to pursue an alternative course of action.

**Comment by BarbossHack** [comment]
Advantage of Json is that you don't have to add all unnecessary fields to make it compatible, just adding the versions under the same tree is sufficient.

Thank you for your answer.

**Comment by BarbossHack** [comment]
Or at least could you share your api server code on github ?
So we will be able to deploy our own server (for offline uses for example).

Thank you.

**Comment by serayuzgur** [comment]
> Or at least could you share your api server code on github ? So we will be able to deploy our own server (for offline uses for example).
> 
> Thank you.

What is the rationale behind requiring offline usage if you're unable to access the latest version information and install the desired version of the crate you seek? 

I am planning some scenarios for the server , In progress.

**Comment by pingufreak** [comment]
Why is the domain hosted in Island while you are living in Turkey? For me it looks also a bit strange. Identity theft?

**Comment by pingufreak** [comment]
> Okay, here's the whole deal: At first, we were using the crates.io API, but they asked us not to and switch to the GitHub API instead.
May I ask you who declined to use crates.io? Namecheap is blocked in a lot of security appliances. For instance at PaloAlto the url https://api.crates-vsc.space/ is categorized as insufficient-content. If it would be possible to get it back to crates.io, it would be very nice. 

Thank you

**Comment by serayuzgur** [comment]
@pingufreak 
> May I ask you who declined to use crates.io?

I can't recall the individual responsible for managing the crates.io API.  3-4 years ago. 

> Namecheap is blocked in a lot of security appliances.
I can  transfer domain to Digital Ocean. Will it function properly?

> the url https://api.crates-vsc.space/ is categorized as insufficient-content
Maybe you can help me to make url trustable.  I don't understand why it is considered like that.

**Comment by pingufreak** [comment]
Why is this closed? If you can change it back to crates.io it would be the best. :)

**Comment by BarbossHack** [comment]
> Why is this closed? If you can change it back to crates.io it would be the best. :)

It was closed because he didn't wanted to uses crates.io, mainly for optimisation reasons (see https://github.com/serayuzgur/crates/issues/195#issuecomment-1640801531) (imho the only valid reason)

**Comment by serayuzgur** [comment]
> > Why is this closed? If you can change it back to crates.io it would be the best. :)
> 
> It was closed because he didn't wanted to uses crates.io, mainly for optimisation reasons (see [#195 (comment)](https://github.com/serayuzgur/crates/issues/195#issuecomment-1640801531)) (imho the only valid reason)

In addition several other minor issues that do not affect your life but do affect mine 😄 . Base for  bigger features like  private registry support, on premise crates index server, etc.

**Comment by WUST-mengqinyu** [comment]
But I have problem with this host. Is there any alternative？
`Completed with errors xxx: Error: statusCode=504`

**Comment by junderw** [comment]
Just thought I'd link to a discussion regarding this issue.

https://www.reddit.com/r/rust/comments/15ye9he/i_forked_crates_vscode_extension_to_make_it_use/

Seems as though a fork of this project is using the sparse HTTP protocol to get the info directly from crates.io's AWS S3 files directly (which is much easier on them because it doesn't even touch their API)

Would be nice if that could be merged. Not that I don't trust the new author, but it seems like post-sparse-protocol this is the optimal method for doing this.

**Comment by BarbossHack** [comment]
> Would be nice if that could be merged. Not that I don't trust the new author, but it seems like post-sparse-protocol this is the optimal method for doing this.

(Don't worry you're totally right, and if this get merged I would even delete my fork :+1: )

**Comment by junderw** [comment]
https://github.com/serayuzgur/crates/compare/84b70c44df439ae83633a6197c67214296542afb...BarbossHack:crates-io:2ab0c2f649b9037ca2e3717a0592eac1dfecb94b

The diff seems pretty small. Perhaps preparing a PR for this repo (without all the funding removals and renamings etc.) would be appropriate.

**Comment by BarbossHack** [comment]
see https://github.com/serayuzgur/crates/pull/204

I know that @serayuzgur has other reasons not to use crates.io (whether with the api or sparse index, see https://github.com/serayuzgur/crates/issues/195#issuecomment-1640801531 and https://github.com/serayuzgur/crates/issues/197#issuecomment-1671600461), so I'm quite sure this PR will be closed, but at least I tried

**Comment by serayuzgur** [comment]
@simonsan  I will think about it. I need to calculate the effect of this with new planned features

**Comment by serayuzgur** [comment]
Since the PR is replacing the code but not implementing a different api connection code regarding to configuration I can not accept it. @BarbossHack  Thanks for your effort, please take a look at my comments on PR

**Comment by serayuzgur** [comment]
@BarbossHack  I'd kindly request that you consider removing it from the market altogether. Given that the fork no longer offers offline support due to the shift towards a sparse approach, it might inadvertently mislead users who are seeking that functionality. This step would ensure that users make informed decisions and that we avoid any potential confusion.

**Comment by serayuzgur** [comment]
> As a long-time user of `crates-io` extension, I would appreciate, if the fork would stay up at least until this issue is resolved and the PR (with the additional feature of choosing the index url in the settings) is merged. When I read about this issue here, I immediately switched to the fork, as I'm in general pretty concerned about my privacy.

@simonsan 
If so , please use older version of crates or use packaged version of the fork. I am not feeling good about it all.   Regarding privacy I'm making an effort to provide reasonable justifications, but maintaining my stance is becoming more difficult. Please explain how the crate name being the only information transmitted to the server can cause you to be concerned about privacy.

**Comment by BarbossHack** [comment]
> @BarbossHack  I'd kindly request that you consider removing it from the market altogether. Given that the fork no longer offers offline support due to the shift towards a sparse approach, it might inadvertently mislead users who are seeking that functionality. This step would ensure that users make informed decisions and that we avoid any potential confusion.

I am currently using my fork offline, so it still allow offline compatibility for users. I have a sparse index clone offline (git clone) with fake dns, so crates-io is able to fetch it (as it would with crates.io api, I would have used the same process). Maybe it wasn't clear, but when I mean "offline" I mean "on a network that does not have internet connection", not simply "local fetch"

I saw your comments on PR, I will be able to make the changes soon (can't do it at work), thank you for allowing us to implement this feature 👍

**Comment by serayuzgur** [comment]
published v 0.6.2

**Comment by serayuzgur** [comment]
@BarbossHack  please check changelog
